src/main/java/com/kobylynskyi/graphql/codegen/mapper/MapperUtils.java

```diff
@@ -46,30 +47,51 @@ static String capitalizeIfRestricted(String fieldName) {
      * @param document      Parent GraphQL document
      * @return Names of all unions that requested <code>definition</code> is part of.
      */
-    static List<String> getUnionsHavingType(MappingConfig mappingConfig,
-                                            NamedNode definition,
-                                            Document document) {
+    static List<String> getUnionNamesHavingType(MappingConfig mappingConfig,
+                                                NamedNode<?> definition,
+                                                Document document) {
         return document.getDefinitions().stream()
-                .filter(def -> def instanceof UnionTypeDefinition)
+                // should cover extensions too
+                .filter(def -> UnionTypeDefinition.class.isAssignableFrom(def.getClass()))
                 .map(def -> (UnionTypeDefinition) def)
                 .filter(union -> isDefinitionPartOfUnion(definition, union))
                 .map(UnionTypeDefinition::getName)
                 .map(unionName -> getClassNameWithPrefixAndSuffix(mappingConfig, unionName))
                 .collect(Collectors.toList());
     }
 
+    /**
+     * Returns a list of definitions of the specific type and name.
+     * It uses {@link java.lang.Class#isAssignableFrom(Class)} for the test
+     *
+     * @param definitionClass the definition class
+     * @param definitionName  the definition name
+     * @param <T>             the type of definition
+     * @return a list of definitions of that class/name or empty list
+     */
+    static <T extends Definition<?>> List<T> getDefinitionsOfType(Document document,
+                                                                  Class<T> definitionClass,
+                                                                  String definitionName) {
+        return document.getDefinitions().stream()
```

Comment:
Wait, I thought we were saving on `getDefinitions()` iterations for the sake of performance, but now we're iterating many times over it here.

I believe we should instead process the whole document once and build an intermediate representation that suits our needs, and then use this intermediate representation to generate classes/interfaces.
This way it would clearly separate the code processing the schema (which may need to change when `graphql-jaav` is updated) and the code building the data model for the generation.

Reply:
We could even define only one such class with generic types:

```java
class ExtendedDefinition<B, E> {
    private B baseDefinition;
    private Collection<E> extensions;

    // getters/setters/constructors
}
```

Reply:
Thanks for your feedback. You are right.
But in order to make this PR clear, let me do this performance improvement in the next PR.
Thank you!

Reply:
Definitely.